### PR TITLE
📖 book: add v1.1=>v1.2 migration doc

### PR DESCRIPTION
**Pull request by sbueringer**
Signed-off-by: Stefan Büringer buringerst@vmware.com

**What this PR does / why we need it**:
This PR adds the initial migration doc for v1.1=>v1.2.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Related #5967

### CODE REVIEW

**Review by killianmuldoon — APPROVED** [review]
/lgtm

Should we maintain a 0.4 => 1.1/1.2 version as well? this is the more likely upgrade path. The current structure is confusing to me as it mixes api changes with code version changes. I think we should find a way to express those seperately.

**Comment by sbueringer** [comment]
> /lgtm
> 
> Should we maintain a 0.4 => 1.1/1.2 version as well? this is the more likely upgrade path. The current structure is confusing to me as it mixes api changes with code version changes. I think we should find a way to express those seperately.

I think it's fine. This documentation is targeted mostly to infrastructure provider authors. I think it's better to document incremental changes between each CAPI release instead of also (redundantly (?)) documenting v0.4=>v1.1. (If I understood you correctly)

**Comment by sbueringer** [comment]
Forgot to add it to the summary. That happens when I think once I don't have to verify a book change with make server-book :/

**Comment by killianmuldoon** [comment]
There's the inconsistency here which I think is confusing:
```
        - [v1alpha3 to v1alpha4](./developer/providers/v1alpha3-to-v1alpha4.md)
        - [v1alpha4 to v1beta1](./developer/providers/v1alpha4-to-v1beta1.md)
        - [v1.0 to v1.1](./developer/providers/v1.0-to-v1.1.md)
        - [v1.1 to v1.2](./developer/providers/v1.1-to-v1.2.md) 
 ```

Should we change the older migration docs to their code versions (which are equivalent to the api versions anyway)? Or should we break these off into different sections? Right now it make it hard to follow (at a glance at least)

**Comment by sbueringer** [comment]
> There's the inconsistency here which I think is confusing:
> 
> ```
>         - [v1alpha3 to v1alpha4](./developer/providers/v1alpha3-to-v1alpha4.md)
>         - [v1alpha4 to v1beta1](./developer/providers/v1alpha4-to-v1beta1.md)
>         - [v1.0 to v1.1](./developer/providers/v1.0-to-v1.1.md)
>         - [v1.1 to v1.2](./developer/providers/v1.1-to-v1.2.md) 
> ```
> 
> Should we change the older migration docs to their code versions (which are equivalent to the api versions anyway)? Or should we break these off into different sections? Right now it make it hard to follow (at a glance at least)

Ah now I got it. Yeah the problem is that we previously used the api contract as a synonym for the version. I'll change it to the release versions. Let's see if someone objects

**Comment by sbueringer** [comment]
Thx good point! Fixed now.

**Comment by killianmuldoon** [comment]
Thanks!

/lgtm

File: docs/book/src/SUMMARY.md [hunk 1]
```diff
@@ -58,11 +58,12 @@
         - [Multi-tenancy](./developer/architecture/controllers/multi-tenancy.md)
         - [Support multiple instances](./developer/architecture/controllers/support-multiple-instances.md)
     - [Provider Implementers](./developer/providers/implementers.md)
-        - [v1alpha1 to v1alpha2](./developer/providers/v1alpha1-to-v1alpha2.md)
-        - [v1alpha2 to v1alpha3](./developer/providers/v1alpha2-to-v1alpha3.md)
-        - [v1alpha3 to v1alpha4](./developer/providers/v1alpha3-to-v1alpha4.md)
-        - [v1alpha4 to v1beta1](./developer/providers/v1alpha4-to-v1beta1.md)
+        - [v0.1 to v0.2](./developer/providers/v0.1-to-v0.2.md)
+        - [v0.2 to v0.3](./developer/providers/v0.2-to-v0.3.md)
+        - [v0.3 to v0.4](./developer/providers/v0.3-to-v0.4.md)
```
Comment (fabriziopandini) [inline docs/book/src/SUMMARY.md:63]: Q: Given that we are preserving & linking older versions of the books, should we start dropping old migration documents and keep only the latest 2/3 (for sake of simplicity)?

**Review by fabriziopandini — COMMENTED** [review]
Only one question, otherwise lgtm

File: docs/book/src/SUMMARY.md (hunk 1 above)
Comment (sbueringer) [reply docs/book/src/SUMMARY.md]: Good point. I think it's fine to only keep the last 2/3 around. I kept everything v0.4+ for now given that we currently don't have links for a v1.0/v1.1 book (pending discussion on the book issue)

**Comment by sbueringer** [comment]
@fabriziopandini PTAL

File: docs/book/src/developer/providers/v1alpha1-to-v1alpha2.md [hunk 2]
```diff
@@ -1,66 +1 @@
...
-running management cluster in whatever environment the provider was built for. For example, Cluster-API-Provider-AWS
-packaged a `clusterctl` that created a Kubernetes cluster in EC2 and installed the necessary controllers to respond to
-Cluster API's APIs.
-
-### v1alpha2
-
-`clusterctl` is likely becoming provider-agnostic meaning one clusterctl is bundled with Cluster API and can be reused
-across providers. Work here is still being figured out but providers will not be packaging their own `clusterctl`
-anymore.
+# v1alpha1 to v1alpha2
```
Comment (fabriziopandini) [inline docs/book/src/developer/providers/v1alpha1-to-v1alpha2.md:1]: Should this be deleted? (It is not linked anymore in the summary)

File: docs/book/src/developer/providers/v1alpha2-to-v1alpha3.md [hunk 3]
```diff
@@ -1,515 +1 @@
...
-> In the future, we'll revisit this user experience and provide a centralized way to configure gates across all Cluster API (inc. providers) controllers.
-
-## clusterctl
-
-`clusterctl` is now bundled with Cluster API, provider-agnostic and can be reused across providers.
-It is the recommended way to setup a management cluster and it implements best practices to avoid common mis-configurations
-and for managing the life-cycle of deployed providers, e.g. upgrades.
-
-see [clusterctl provider contract](../../clusterctl/provider-contract.md) for more details.
+# v1alpha2 to v1alpha3
```
Comment (fabriziopandini) [inline docs/book/src/developer/providers/v1alpha2-to-v1alpha3.md:1]: Should this be deleted? (It is not linked anymore in the summary)

File: docs/book/src/developer/providers/v1alpha3-to-v1alpha4.md [hunk 4]
```diff
@@ -1,358 +1 @@
...
-## Required spec fields for Control Planes using 'Machines' for ClusterClass support
-ControlPlane implementations that use an underlying MachineInfrastructure must now [include a 'machineTemplate' as defined below, with subordinate fields, in its Spec.](../architecture/controllers/control-plane.md#required-spec-fields-for-implementations-using-machines)
-
-
- `machineTemplate` - is a struct containing details of the control plane
-  machine template.
-
- Please note that implementing this field for control plane providers using machines is a requirement for ClusterClass and managed topologies to work consistently across all providers.
-
+# v1alpha3 to v1alpha4
```
Comment (fabriziopandini) [inline docs/book/src/developer/providers/v1alpha3-to-v1alpha4.md:1]: Should this be deleted? (It is not linked anymore in the summary)

File: docs/book/src/developer/providers/v1alpha4-to-v1beta1.md [hunk 5]
```diff
@@ -1,73 +1 @@
...
-This has no user facing impact on brand-new clusters created as v1beta1.
-For Cluster API running clusters upgraded through clusterctl this should be ok given that we stop the old controllers.
-Users relying on custom upgrades procedures should ensure a migration to v1alpha4 (multilock "configmapsleases") first, which will acquire a leader lock on both resources. After that, they can proceed migrating to v1beta1 ("leases"). As an additional safety step, these users should ensure the old controllers are stopped before running the new ones with the new lock mechanism.
-Otherwise, your controller might end up with multiple running instances that each acquired leadership through different resource locks during upgrades and thus act on the same resources concurrently.
-
-
-[cluster-contract]: ./cluster-infrastructure.md
-[machine-contract]: ./machine-infrastructure.md
-[capv-ipam]: https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/pull/1210
+# v1alpha4 to v1beta1
```
Comment (fabriziopandini) [inline docs/book/src/developer/providers/v1alpha4-to-v1beta1.md:1]: Should this be deleted? (replaced by v0.4-to-v1.0.md)

File: docs/book/src/tasks/upgrading-cluster-api-versions.md [hunk 6]
```diff
@@ -35,5 +35,5 @@ You should now be able to manage your resources using the `v1alpha4` version of
 <!-- links -->
 [components]: ../reference/glossary.md#provider-components
 [management cluster]: ../reference/glossary.md#management-cluster
-[Cluster API v1alpha3 compared to v1alpha4 section]: ../developer/providers/v1alpha3-to-v1alpha4.md
-[Cluster API v1alpha4 compared to v1beta1 section]: ../developer/providers/v1alpha4-to-v1beta1.md
+[Cluster API v1alpha3 compared to v1alpha4 section]: ../developer/providers/v0.3-to-v0.4.md
+[Cluster API v1alpha4 compared to v1beta1 section]: ../developer/providers/v0.4-to-v1.0.md
```
Comment (fabriziopandini) [inline docs/book/src/tasks/upgrading-cluster-api-versions.md:39]: should this be
```suggestion
[Cluster API v0.4 compared to v1.0 section]: ../developer/providers/v0.4-to-v1.0.md
```

**Comment by sbueringer** [comment]
Should be fixed now

**Review by vincepri — APPROVED** [review]
/approve
/lgtm
